EthProvider.go

```diff
@@ -0,0 +1,131 @@
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/pkg/errors"
+	"github.com/zksync-sdk/zksync-go/contracts/ERC20"
+	"github.com/zksync-sdk/zksync-go/contracts/ZkSync"
+	"math/big"
+)
+
+type EthProvider interface {
+	ApproveDeposits(token *Token, limit *big.Int, options *GasOptions) (*types.Transaction, error)
+	IsDepositApproved(token *Token, userAddress common.Address, threshold *big.Int) (bool, error)
+	Deposit(token *Token, amount *big.Int, userAddress common.Address, options *GasOptions) (*types.Transaction, error)
+	SetAuthPubkeyHash(pubKeyHash string, zkNonce uint32, options *GasOptions) (*types.Transaction, error)
+	IsOnChainAuthPubkeyHashSet(nonce uint32) (bool, error)
+	GetBalance() (*big.Int, error)
+	GetNonce() (uint64, error)
+	FullExit(token *Token, accountId uint32, options *GasOptions) (*types.Transaction, error)
+}
+
+type DefaultEthProvider struct {
+	client   *ethclient.Client
+	contract *ZkSync.ZkSync
+	address  common.Address
+	auth     *bind.TransactOpts
+}
+
+type GasOptions struct {
+	GasPrice *big.Int // Gas price to use for the transaction execution (nil = gas price oracle)
+	GasLimit uint64   // Gas limit to set for the transaction execution (0 = estimate)
+}
+
+func (p *DefaultEthProvider) ApproveDeposits(token *Token, limit *big.Int, options *GasOptions) (*types.Transaction, error) {
+	tokenContract, err := ERC20.NewERC20(token.GetAddress(), p.client)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to load token contract")
+	}
+	if limit == nil {
+		// max approve amount 2^256 - 1
+		limit = big.NewInt(0).Sub(big.NewInt(0).Exp(big.NewInt(2), big.NewInt(256), nil), big.NewInt(1))
+	}
+	auth := p.getAuth(options)
+	return tokenContract.Approve(auth, p.address, limit)
+}
+
+func (p *DefaultEthProvider) IsDepositApproved(token *Token, userAddress common.Address, threshold *big.Int) (bool, error) {
+	tokenContract, err := ERC20.NewERC20(token.GetAddress(), p.client)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to load token contract")
+	}
+	auth := &bind.CallOpts{}
+	allowed, err := tokenContract.Allowance(auth, userAddress, p.address)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to call Allowance view of token contract")
+	}
+	if threshold == nil {
+		// default threshold 2^255
+		threshold = big.NewInt(0).Exp(big.NewInt(2), big.NewInt(255), nil)
+	}
+	return allowed.Cmp(threshold) >= 0, nil
+}
+
+func (p *DefaultEthProvider) Deposit(token *Token, amount *big.Int, userAddress common.Address, options *GasOptions) (*types.Transaction, error) {
+	auth := p.getAuth(options)
+	auth.Value = amount
+	if token.IsETH() {
+		return p.contract.DepositETH(auth, userAddress)
+	} else {
+		return p.contract.DepositERC20(auth, token.GetAddress(), amount, userAddress)
+	}
+}
+
+func (p *DefaultEthProvider) Transfer(token *Token, amount *big.Int, recipient common.Address, options *GasOptions) (*types.Transaction, error) {
```

Comment:
It is present in Java and Rust SDK .. But I'll be glad to remove it because I couldn't find a simple wrapper for ETH transfers, only for ERC20 tokens.

Reply:
I see. Plain erc20 transfers are out of scope of our sdks